pkg/broker/broker.go

```diff
@@ -850,10 +859,194 @@ func (a AnsibleBroker) Unbind(
+	}
+
+	// Make sure that we can find the plan we're updating from. This should probably never fail, but cover our tail.
+	var updateFromPlanKey = -1
+	for k, v := range spec.Plans {
+		if v.Name == (*si.Parameters)["_apb_plan_id"] {
+			updateFromPlanKey = k
+		}
+	}
+	if updateFromPlanKey == -1 {
+		a.log.Error("The plan %s, specified for updating from on instance %s, does not exist.", (*si.Parameters)["_apb_plan_id"], si.ID)
+		return nil, ErrorPlanNotFound
+	}
+	updateFromPlan := spec.Plans[updateFromPlanKey]
+
+	// Make sure we can find the plan we're updating to.
+	var updateToPlanKey = -1
+	for k, v := range spec.Plans {
+		if v.Name == req.PlanID {
+			updateToPlanKey = k
+		}
+	}
+	if updateToPlanKey == -1 {
+		a.log.Error("The plan %s, specified for updating to on instance %s, does not exist.", req.PlanID, si.ID)
+		return nil, ErrorPlanNotFound
+	}
+	updateToPlan := spec.Plans[updateFromPlanKey]
+
+	// Make sure that the current plan supports updating to the requested plan if a plan update was requested.
+	if req.PlanID != (*si.Parameters)["_apb_plan_id"] {
+		var updatable = false
+		for _, v := range updateFromPlan.UpdatesTo {
+			if v == req.PlanID {
+				updatable = true
+				(*si.Parameters)["_apb_plan_id"] = req.PlanID
+			}
+		}
+		if updatable == false {
+			a.log.Error("The current plan, %s, cannot be updated to the requested plan, %s.", (*si.Parameters)["_apb_plan_id"], req.PlanID)
+			return nil, ErrorPlanUpdateNotPossible
+		}
+	}
+
+	// Make sure that parameters requested for update exist on the new plan and that they are updatable.
+	for k, v := range req.Parameters {
+		var paramUpdated = false
+		for _, key := range updateToPlan.Parameters {
+			if key.Name == k {
+				if key.Updatable == false {
+					a.log.Error("Tried to update non-updatable parameter, %s, on instance %s.", k, si.ID)
+					return nil, ErrorParameterNotUpdatable
+				}
+				(*si.Parameters)[k] = v
+				paramUpdated = true
+			}
+		}
+		if paramUpdated == false {
+			a.log.Error("Parameter %s, requested for update on instance %s, does not exist.", k, si.ID)
+			return nil, ErrorParameterNotFound
+		}
+	}
+
+	// We're ready to provision so save
+	if err = a.dao.SetServiceInstance(instanceUUID.String(), si); err != nil {
+		return nil, err
+	}
+
+	var token string
+
+	if async {
+		a.log.Info("ASYNC update in progress")
+		// asyncronously provision and return the token for the lastoperation
+		pjob := NewProvisionJob("update", si, a.clusterConfig, a.log)
+
+		token, err = a.engine.StartNewJob("", pjob, ProvisionTopic)
+		if err != nil {
+			a.log.Error("Failed to start new job for async update\n%s", err.Error())
+			return nil, err
+		}
+
+		// HACK: there might be a delay between the first time the state in etcd
+		// is set and the job was already started. But I need the token.
+		a.dao.SetState(instanceUUID.String(), apb.JobState{Token: token, State: apb.StateInProgress})
+	} else {
+		// TODO: do we want to do synchronous updating?
```

Comment:
We'll likely only see async requests for update from the catalog, but we should probably support sync updates. If not, we should explicitly 422.